File: 00-Meta/Scripts/vault-cli.py
#!/usr/bin/env python3
import sqlite3
import sys
from pathlib import Path

VAULT = Path(__file__).parent
DB = VAULT / ".vault-index.db"

def search(query, limit=10):
    conn = sqlite3.connect(DB)
    c = conn.cursor()
    c.execute('''SELECT path, title FROM fts_notes 
                 WHERE fts_notes MATCH ? ORDER BY rank LIMIT ?''', (query, limit))
    results = c.fetchall()
    conn.close()
    
    print(f"\n🔍 Search results for '{query}':\n")
    for i, (path, title) in enumerate(results, 1):
        print(f"{i}. {title}")
        print(f"   📁 {path}\n")

def stats():
    conn = sqlite3.connect(DB)
    c = conn.cursor()
    
    c.execute("SELECT COUNT(*) FROM notes")
    notes = c.fetchone()[0]
    c.execute("SELECT COUNT(*) FROM links")
    links = c.fetchone()[0]
    c.execute("SELECT COUNT(*) FROM tags")
    tags = c.fetchone()[0]
    c.execute("SELECT COUNT(*) FROM formulas")
    formulas = c.fetchone()[0]
    
    c.execute("SELECT subject, COUNT(*) FROM notes WHERE subject IS NOT NULL GROUP BY subject")
    by_subject = c.fetchall()
    
    c.execute("SELECT tag, count FROM tags ORDER BY count DESC LIMIT 10")
    top_tags = c.fetchall()
    
    conn.close()
    
    print("\n📊 Vault Statistics:\n")
    print(f"  Total Notes: {notes}")
    print(f"  Total Links: {links}")
    print(f"  Total Tags: {tags}")
    print(f"  Total Formulas: {formulas}")
    
    print("\n📚 Notes by Subject:")
    for subj, count in by_subject:
        print(f"  {subj}: {count}")
    
    print("\n🏷️  Top Tags:")
    for tag, count in top_tags:
        print(f"  #{tag}: {count}")

def orphans():
    conn = sqlite3.connect(DB)
    c = conn.cursor()
    c.execute('''SELECT path, title FROM notes 
                 WHERE path NOT IN (SELECT DISTINCT source_path FROM links)
                 AND path NOT IN (SELECT DISTINCT target_path FROM links)
                 AND type != 'moc' ''')
    results = c.fetchall()
    conn.close()
    
    print(f"\n🔗 Orphaned Notes ({len(results)}):\n")
    for path, title in results:
        print(f"  • {title} ({path})")

def formulas_search(query):
    conn = sqlite3.connect(DB)
    c = conn.cursor()
    c.execute('''SELECT DISTINCT path, formula FROM formulas 
                 WHERE formula LIKE ? LIMIT 20''', (f'%{query}%',))
    results = c.fetchall()
    conn.close()
    
    print(f"\n📐 Formulas containing '{query}':\n")
    for path, formula in results:
        print(f"  {formula}")
        print(f"  📁 {path}\n")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: ./vault-cli.py <command> [args]")
        print("\nCommands:")
        print("  search <query>    - Search notes")
        print("  stats             - Show vault statistics")
        print("  orphans           - Find unlinked notes")
        print("  formulas <query>  - Search formulas")
        sys.exit(1)
    
    cmd = sys.argv[1]
    
    if cmd == "search" and len(sys.argv) > 2:
        search(' '.join(sys.argv[2:]))
    elif cmd == "stats":
        stats()
    elif cmd == "orphans":
        orphans()
    elif cmd == "formulas" and len(sys.argv) > 2:
        formulas_search(' '.join(sys.argv[2:]))
    else:
        print(f"Unknown command: {cmd}")

File: 00-Meta/src/vault_cli.rs
use anyhow::Result;
use clap::{Parser, Subcommand};
use rusqlite::Connection;
use serde_json::json;

#[derive(Parser)]
#[command(name = "vault-cli")]
#[command(about = "Study Vault CLI - Search and manage your notes")]
struct Cli {
    #[command(subcommand)]
    command: Commands,
}

#[derive(Subcommand)]
enum Commands {
    /// Search notes
    Search {
        query: String,
        #[arg(short, long, default_value = "10")]
        limit: usize,
    },
    /// Show vault statistics
    Stats,
    /// Find orphaned notes
    Orphans,
    /// Search formulas
    Formulas {
        query: String,
        #[arg(short, long, default_value = "10")]
        limit: usize,
    },
}

fn main() -> Result<()> {
    let cli = Cli::parse();
    let db_path = ".vault-index.db";
    
    match cli.command {
        Commands::Search { query, limit } => search(&query, limit, db_path)?,
        Commands::Stats => stats(db_path)?,
        Commands::Orphans => orphans(db_path)?,
        Commands::Formulas { query, limit } => formulas(&query, limit, db_path)?,
    }
    
    Ok(())
}

fn search(query: &str, limit: usize, db_path: &str) -> Result<()> {
    let conn = Connection::open(db_path)?;
    let mut stmt = conn.prepare(
        "SELECT path, title FROM fts_notes 
         WHERE fts_notes MATCH ? ORDER BY rank LIMIT ?"
    )?;
    
    let results: Vec<(String, String)> = stmt
        .query_map([query, &limit.to_string()], |row| {
            Ok((row.get(0)?, row.get(1)?))
        })?
        .collect::<Result<Vec<_>, _>>()?;
    
    println!("\n🔍 Search results for '{}':\n", query);
    for (i, (path, title)) in results.iter().enumerate() {
        println!("{}. {}", i + 1, title);
        println!("   📁 {}\n", path);
    }
    
    Ok(())
}

fn stats(db_path: &str) -> Result<()> {
    let conn = Connection::open(db_path)?;
    
    let notes: i64 = conn.query_row("SELECT COUNT(*) FROM notes", [], |row| row.get(0))?;
    let links: i64 = conn.query_row("SELECT COUNT(*) FROM links", [], |row| row.get(0))?;
    let tags: i64 = conn.query_row("SELECT COUNT(*) FROM tags", [], |row| row.get(0))?;
    let formulas: i64 = conn.query_row("SELECT COUNT(*) FROM formulas", [], |row| row.get(0))?;
    
    let mut stmt = conn.prepare("SELECT subject, COUNT(*) FROM notes WHERE subject IS NOT NULL GROUP BY subject")?;
    let by_subject: Vec<(String, i64)> = stmt
        .query_map([], |row| Ok((row.get(0)?, row.get(1)?)))?
        .collect::<Result<Vec<_>, _>>()?;
    
    println!("\n📊 Vault Statistics:\n");
    println!("  Total Notes: {}", notes);
    println!("  Total Links: {}", links);
    println!("  Total Tags: {}", tags);
    println!("  Total Formulas: {}", formulas);
    
    println!("\n📚 Notes by Subject:");
    for (subject, count) in by_subject {
        println!("  {}: {}", subject, count);
    }
    
    Ok(())
}

fn orphans(db_path: &str) -> Result<()> {
    let conn = Connection::open(db_path)?;
    let mut stmt = conn.prepare(
        "SELECT path, title FROM notes 
         WHERE path NOT IN (SELECT DISTINCT source_path FROM links)
         AND path NOT IN (SELECT DISTINCT target_path FROM links)
         AND type != 'moc'"
    )?;
    
    let orphans: Vec<(String, String)> = stmt
        .query_map([], |row| Ok((row.get(0)?, row.get(1)?)))?
        .collect::<Result<Vec<_>, _>>()?;
    
    println!("\n🔗 Orphaned Notes ({}):\n", orphans.len());
    for (path, title) in orphans {
        println!("  • {} ({})", title, path);
    }
    
    Ok(())
}

fn formulas(query: &str, limit: usize, db_path: &str) -> Result<()> {
    let conn = Connection::open(db_path)?;
    let mut stmt = conn.prepare(
        "SELECT DISTINCT path, formula FROM formulas 
         WHERE formula LIKE ? LIMIT ?"
    )?;
    
    let pattern = format!("%{}%", query);
    let results: Vec<(String, String)> = stmt
        .query_map([&pattern, &limit.to_string()], |row| {
            Ok((row.get(0)?, row.get(1)?))
        })?
        .collect::<Result<Vec<_>, _>>()?;
    
    println!("\n📐 Formulas containing '{}':\n", query);
    for (path, formula) in results {
        println!("  {}", formula);
        println!("  📁 {}\n", path);
    }
    
    Ok(())
}
